### ISSUE TITLE
Planned Cost should use new costing model

### ISSUE THREAD

**Issue by ianhinder** [UoMResearchIT/CapX#332]
As discussed in the PSM meeting on 26-Jun-2024, CapX has a Planned Cost field for a project which is calculated based on the day rates assigned to RSEs in the project's tasks. This does not reflect the new costing model, which includes a management fee (charged whether a task is running or not, I believe).

CapX should provide a Planned Cost based on the new costing model.  As noted in the meeting, an initial version of this might use the day rates for the first year we have been given from BlackDackel, as it is complicated to add in all the increases that BlackDackel will have assumed for subsequent years. This means that the CapX Planned Cost will be lower than the budget for multi-year projects.

Without an accurate Planned Cost, it is very hard to juggle around details of a project such as bringing in different RSEs or introducing pauses etc.

**Comment by SJaffa** [comment]
I was just playing with CapX and thinking this - it would be great to have PMs assigned to projects if there is a way to tag roles on tasks (dev/PM) so that people can still have their %FTE for different roles.

**Comment by aharwood2** [comment]
Dependent on #308 and to some extent #98.

**Comment by aharwood2** [comment]
- [x] Create a new table and CRUD interface for maintaining reference costs in the database for a given financial year.
- [x] Add Model A / B / C options and enable day rate field only when applicable.
- [x] Add actual hours per resource rather than collectively.
- [x] Update the cost calculations to use the financial references where possible and where choice of model requires it.
- [x] Ensure planned hours, planned costs, actual hours and actual costs work collectively across Project, SubTask and Resource models -- need a SQL script to distribute the actual hours across the resources in proportion and calculate costs based on day rate given.
- [x] Have leadership planned and actuals as separate field so they can be viewed separate to the Planned and Actual cost fields -- maybe as a popup if you hover over an info icon?

**Comment by ianhinder** [comment]
- We will need to keep the old day rate functionality, so that projects with an agreed day rate are still valid.
- The salary costs increase year on year. We could try to find out the algorithm used by BlackDackel and copy it.
- The increases are presumably only modelled in BlackDackel, since they cannot know how salary costs will increase. Do we go back and update things when the actual salary costs are known?

**Comment by aharwood2** [comment]
I think I'd like to have "model A" and "model B" options so that we can choose which model to use as the new model won't work for legacy projects.

**Comment by aharwood2** [comment]
This is turning out to be incredibly complicated. Given that I don't want to built a tool that a) replicated BlackDackel or b) replicates the financial tracking sheet, what is the purpose of this again? I think I've got lost in the weeds.

I think we want to know:
1. What is the approx cost of the assigned resources to the department? -- Planned Cost
2. What is the approx cost of the work completed (use same approach as 1)? -- Actual Cost

To do this I can:
1. Update the planned cost calculation so that it uses the standard and junior rates as per the memo or the middle of the grade for staff members.
2. Include the leadership charges in actual and planned costs so it aligns with the cost memo.
3. Update the actual cost calculation as per the above.

But:
1. Grade changes mid-task are complicated to include. Possible but is it worth the effort?
2. We only record actuals as a lump value not per FY so 3 in the previous list can't really be done without changing how we record actuals.
3. Also, to include 1 in this list we would also need to record actuals separately for each grade within a financial year.
4. Generally for the last two, we can fix this by essentially building a timesheet system into CapX but this will be a lot of work.

**Comment by rowleya** [comment]
I would say:
 1. Leave the ability to set the cost of the task as it is now.  This might be needed when the final invoice doesn't match up and we have to make the day rate match.
 2. Include a day rate for leadership per task.  This will then allow that to change for different tasks in case we need to split things up between now and future.
 3. Add a default leadership day rate to the project (but still make this changable per task in case required).
 
With those things we can then see the total cost of the project overall including management charges.  The defaults will be the standard day rates, though these can be changed on the project anyway.  If there is a grade change mid-project and this needs to be accounted for, simply stop the old task and create a new one with the new rates.  I think that makes it easy enough to implement and not too hard to use...

**Comment by aharwood2** [comment]
Thanks @rowleya. Why would we need to set a day rate for leadership? It is always the same isn't it? I'm conscious I don't want to add more knobs when we don't need them!

**Comment by aharwood2** [comment]
I've opted to keep the costing simple so there are two restrictions:
1. Mid-task grade changes are ignored.
2. If using salary costs rather than a day rate, the salary costs are assumed to remain constant through the task (not true for multi-year tasks but I think this is OK as an approximation).

Both these things could be relaxed in the future but we would have to record actuals weekly so we can capture which hours booked fall under which grade and which financial year etc. At the moment we don't have that ability so we can't make this more accurate than this.

**Comment by rowleya** [comment]
> Thanks @rowleya. Why would we need to set a day rate for leadership? It is always the same isn't it? I'm conscious I don't want to add more knobs when we don't need them!

The only reason I suggested this was similar to the issue of the day rate for the RSE, i.e. it might change mid project, so it will be easier to have a split task if that does happen.  I can see it changing every year, so it could be useful on multi-year projects.  It also likely means you just copy everything that has been done for the day rate of the task into the management day rate of the task, so it will make the code potentially quite uniform too ;).

> Both these things could be relaxed in the future but we would have to record actuals weekly so we can capture which hours booked fall under which grade and which financial year etc. At the moment we don't have that ability so we can't make this more accurate than this.

If there is a mid-task "grade" change (or just that the project spans enough time that the day rate changes), it just means you split the task and then have two separate day rates.  Hours are recorded weekly in the timesheets anyway, so it isn't that hard to split the actuals also (I have already had to do that when the RSE assignment changes between 0.2 and 0.6 in one task for example, and it isn't that hard to keep track of).
